runtime/bindings/python/loop.cc

```diff
@@ -0,0 +1,254 @@
+        py::gil_scoped_acquire acquire_gil;
+        CheckApiStatus(status,
+                       "iree_hal_device_wait_semaphores from HalDeviceLoop");
+      }
+
+      status = iree_hal_semaphore_query(control_sem_, &next_control_wakeup);
+      if (!iree_status_is_ok(status)) {
+        py::gil_scoped_acquire acquire_gil;
+        CheckApiStatus(status,
+                       "iree_hal_device_wait_semaphores from HalDeviceLoop");
+      }
+      next_control_wakeup += 1;
+    }
+
+    free(wait_semaphores);
+    free(wait_payloads);
+
+    // Cancel all pending futures.
+    {
+      for (auto &entry : pending_futures) {
+        py::handle future = std::get<2>(entry);
+        CancelFuture(future);
+      }
+    }
+  }
+
+  void CancelFuture(py::handle future) {
+    py::gil_scoped_acquire acquire_gil;
+    try {
+      future.attr("cancel")();
+    } catch (py::python_error &e) {
+      ReportUncaughtException(e);
+    }
+    future.dec_ref();
+  }
+
+  void SignalFuture(py::handle future) {
+    py::gil_scoped_acquire acquire_gil;
+    try {
+      future.attr("set_result")(true);
+    } catch (py::python_error &e) {
+      ReportUncaughtException(e);
+    }
+    future.dec_ref();
+  }
+
+  void SignalFutureFailure(py::handle future, iree_status_t status) {
+    py::gil_scoped_acquire acquire_gil;
+    std::string message = ApiStatusToString(status);
+    iree_status_ignore(status);
+    PyErr_SetString(PyExc_RuntimeError, message.c_str());
+    PyObject *exc_type;
+    PyObject *exc_value;
+    PyObject *exc_tb;
+    PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+    future.attr("set_exception")(exc_value);
+    Py_XDECREF(exc_type);
+    Py_XDECREF(exc_tb);
+    Py_XDECREF(exc_value);
+    future.dec_ref();
+  }
+
+  void SignalShutdown() {
+    iree_slim_mutex_lock(&mu_);
+    shutdown_signalled_ = true;
+    auto status = iree_hal_semaphore_signal(control_sem_, control_next_++);
+    iree_slim_mutex_unlock(&mu_);
+    CheckApiStatus(status, "iree_hal_semaphore_signal");
+  }
+
+  void OnSemaphore(HalSemaphore semaphore, uint64_t payload,
+                   py::handle future) {
+    iree_slim_mutex_lock(&mu_);
+    next_pending_futures_.push_back(
+        std::make_tuple(semaphore.steal_raw_ptr(), payload, std::move(future)));
+    future.inc_ref();
+    auto status = iree_hal_semaphore_signal(control_sem_, control_next_++);
+    iree_slim_mutex_unlock(&mu_);
+    CheckApiStatus(status, "iree_hal_semaphore_signal");
+  }
+
+ private:
+  // Certain calls into Futures may raise exceptions because of illegal states.
+  // There is really not much we can do about this, so we attempt to report.
+  // TODO: Have some kind of fatal exception hook.
+  void ReportUncaughtException(py::python_error &e) {
+    e.discard_as_unraisable(__func__);
+  }
+
+  iree_slim_mutex_t mu_;
+  HalDevice device_;
+  iree_hal_semaphore_t *control_sem_ = nullptr;
+  uint64_t control_next_ = 1;
+  bool shutdown_signalled_ = false;
```

Comment:
nit: `signaled` (`signalled` is british, color/colour :)

Reply:
lol, I think I almost always use British spelling reflexively. Not sure why but probably the same obstinate neural dysfunction that makes me avoid the oxford comma.